src/handlers/register.rs

```diff
@@ -32,24 +34,89 @@ pub async fn handler(
     claims.verify_basic(&state.auth_aud, None)?;
     let client_id = ClientId::from(claims.iss);
 
+    increment_counter!(state.metrics, register);
+
+    if let Some(tags) = body.tags {
+        increment_counter!(state.metrics, registration_overwrite);
+        overwrite_registration(&state, client_id.clone(), tags, body.relay_url).await?;
+    } else {
+        increment_counter!(state.metrics, registration_update);
+        update_registration(
+            &state,
+            client_id.clone(),
+            body.append_tags,
+            body.remove_tags,
+            body.relay_url,
+        )
+        .await?;
+    }
+
+    Ok(Response::default())
+}
+
+async fn overwrite_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    tags: Vec<Arc<str>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
     state
         .registration_store
         .upsert_registration(
             client_id.value(),
-            body.tags.iter().map(AsRef::as_ref).collect(),
-            body.relay_url.as_ref(),
+            tags.iter().map(AsRef::as_ref).collect(),
+            relay_url.as_ref(),
         )
         .await?;
 
     state
         .registration_cache
         .insert(client_id.into_value(), CachedRegistration {
-            tags: body.tags.clone(),
-            relay_url: body.relay_url.clone(),
+            tags,
+            relay_url,
         })
         .await;
 
-    increment_counter!(state.metrics, register);
-
     Ok(Response::default())
 }
+
+async fn update_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    append_tags: Option<Vec<Arc<str>>>,
+    remove_tags: Option<Vec<Arc<str>>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
+    if let (Some(append_tags), Some(remove_tags)) = (append_tags.clone(), remove_tags.clone()) {
+        for tag in remove_tags.iter() {
+            if append_tags.contains(tag) {
+                return Err(Error::InvalidUpdateRequest);
+            }
+        }
```

Comment:
```suggestion
        if remove_tags.intersect(&append_tags).count() > 0 {
            return Err(Error::InvalidUpdateRequest);
        }
```

Reply:
Found a middle ground 😊

Reply:
I think the necessity to collect the iterators between each step feels kinda stupid tho, bad language design 😊

Reply:
Agree, what you did is much nicer!